### PR TITLE
Encode/Decode Implementation

### LINKED ISSUE #2: decode / encode interface
```erlang
smpp:encode(PDU :: map()) -> {ok, PDU :: hex_string()} | {error, Error :: string()}.
```
Example
```erlang
> smpp:encode(#{command_id => 1, ...}).
{ok, "00 AB ..."}
> smpp:encode(#{bad_field => 1, ...}).
{error, "Invalid field bad_field"}
```
---
```erlang
smpp:decode(PDU :: hex_string()) -> {ok, PDU :: map()}.
```
Example
```erlang
> smpp:decode("00 00 AB ...").
{ok, #{command_id => 1, short_message = <<"...">>...}}
> smpp:decode("0000AB...").
{ok, #{command_id => 1, short_message = <<"...">>...}}
> smpp:decode(<<"0000AB...">>).
{ok, #{command_id => 1, short_message = <<"...">>...}}
> smpp:decode("00 00 FF GG ...").
{error, "unable to encode FF GG"}
```

### PR DESCRIPTION
**Pull request by kirankurup**
fixes #2

### CODE REVIEW

File: src/smpp.erl [hunk 1]
```diff
@@ -339,6 +338,37 @@ cmd(cancel_broadcast_sm_resp)   -> ?COMMAND_ID_CANCEL_BROADCAST_SM_RESP;
...
+            lists:flatten(lists:join(" ", [int_to_hex_str(B) || B <- binary_to_list(Bin)]));
+        {error, _, S, _} ->
+            {error, element(3, err(S))}
+    end;
+encode(_) ->
+    {error, "Input to encode should be map"}.
+
+-spec(decode(HEX_STRING :: string()) -> {ok, PDU :: map()} | {error, string()}).
+decode(HexStr) when is_list(HexStr) ->
+    Bin = list_to_binary([binary_to_integer(B, 16) || B <- re:split(HexStr, " ")]),
```
Comment (c-bik) [inline src/smpp.erl]: Will fail for "no space" strings and binaries
image

File: src/smpp.erl [hunk 2]
```diff
@@ -339,6 +338,37 @@ cmd(cancel_broadcast_sm_resp)   -> ?COMMAND_ID_CANCEL_BROADCAST_SM_RESP;
 
 cmd({Cmd,S,SN,B}) -> {cmd(Cmd),S,SN,B}.
 
+-spec(encode(PDU :: map()) -> {ok, HEX_STRING :: string()} | {error, string()}).
+encode(PDU) when is_map(PDU) ->
+    case pack(json2internal(PDU)) of
+        {ok, Bin} ->
+            lists:flatten(lists:join(" ", [int_to_hex_str(B) || B <- binary_to_list(Bin)]));
```
Comment (c-bik) [inline src/smpp.erl]: Do it in lined:
```erlang
> string:join([string:right(integer_to_list(B, 16),2,$0) || <<B>> <= <<0,1,2,3,16,17>>], " ").
"00 01 02 03 10 11"
```

File: src/smpp.erl [hunk 3]
```diff
@@ -112,20 +112,19 @@ pack({CmdId, Status, SeqNum, Body} = SMPP)
...
+            Error;
+        {ok, {CmdId, Status, SeqNum, Body}} ->
+            case lists:member(return_maps, Opts) of
+                true ->
+                    SMPPMap = lists:foldl(fun list_to_map/2, #{}, Body),
+                    SMPPMap#{command_id => CmdId, command_status => Status, sequence_number => SeqNum};
+                _ ->
+                    Hd = [{command_id, CmdId}, {command_status, Status}, {sequence_number, SeqNum}],
+                    Hd ++ lists:foldl(fun list_to_pl/2, [], Body)
+            end
```
Comment (c-bik) [inline src/smpp.erl]: 👍 Thanks.
I, however, would like have it as a separate issue and PR, can you please do that?

File: src/smpp.erl [hunk 4]
```diff
@@ -339,6 +338,37 @@ cmd(cancel_broadcast_sm_resp)   -> ?COMMAND_ID_CANCEL_BROADCAST_SM_RESP;
...
+    case unpack_map(Bin) of
+        {error, _, S, _} ->
+            {error, element(3, err(S))};
+        PDU ->
+            {ok, internal2json(PDU)}
+    end;
+decode(_) ->
+    {error, "Input to decode should be Hex String"}.
+
+int_to_hex_str(0) ->
```
Comment (c-bik) [inline src/smpp.erl]: Can be done in-lined (see above)! No need for an extra function.

File: src/smpp.erl [hunk 5]
```diff
@@ -347,6 +350,42 @@ cmd(cancel_broadcast_sm_resp)   -> ?COMMAND_ID_CANCEL_BROADCAST_SM_RESP;
 
 cmd({Cmd,S,SN,B}) -> {cmd(Cmd),S,SN,B}.
 
+-spec(encode(PDU :: map()) -> {ok, HEX_STRING :: string()} | {error, string()}).
+encode(PDU) when is_map(PDU) ->
+    case pack(json2internal(PDU)) of
+        {ok, Bin} ->
+            {ok, 
+                lists:flatten(string:join([string:pad(integer_to_list(B, 16), 2, leading, $0) || <<B>> <= Bin], " "))};
```
Comment (c-bik) [inline src/smpp.erl]: * why not `string:right`?
* why `lists:flatten` after `string:join`?

File: src/smpp.erl [hunk 6]
```diff
@@ -347,6 +350,42 @@ cmd(cancel_broadcast_sm_resp)   -> ?COMMAND_ID_CANCEL_BROADCAST_SM_RESP;
...
+        {error, _, S, _} ->
+            {error, element(3, err(S))}
+    end;
+encode(_) ->
+    {error, "Input to encode should be map"}.
+
+-spec(decode(HEX_STRING :: string() | binary()) -> {ok, PDU :: map()} | {error, string()}).
+decode(HexStr) when is_list(HexStr) ->
+    HexBin = re:replace(HexStr, " ", "", [global, {return, binary}]),
+    decode(HexBin);
```
Comment (c-bik) [inline src/smpp.erl]: `list_to_binary` should be enough here, as `binary:replace` (in `decode(HexBin) when is_binary(HexBin) ->...`) will take of the space in both cases!

File: src/smpp.erl [hunk 7]
```diff
@@ -347,6 +350,42 @@ cmd(cancel_broadcast_sm_resp)   -> ?COMMAND_ID_CANCEL_BROADCAST_SM_RESP;
...
+decode(HexBin) when is_binary(HexBin) ->
+    ModBin = binary:replace(HexBin, <<" ">>, <<>>, [global]),
+    Bin = list_to_binary(bin_to_int_list(ModBin)),
```
Comment (c-bik) [inline src/smpp.erl]: Use binary comprehension. Default smallest unit is always byte!
Ref : https://github.com/K2InformaticsGmbH/smpp_parser/pull/4/files#diff-eab1ed7e6cbf42527ce0f032ddcf7d61R358

File: src/smpp.erl [hunk 8]
```diff
@@ -347,6 +350,42 @@ cmd(cancel_broadcast_sm_resp)   -> ?COMMAND_ID_CANCEL_BROADCAST_SM_RESP;
...
+    case unpack_map(Bin) of
+        {error, _, S, _} ->
+            {error, element(3, err(S))};
+        PDU ->
+            {ok, internal2json(PDU)}
+    end;    
+decode(_) ->
+    {error, "Input to decode should be Hex String"}.
+
+bin_to_int_list(Bin) ->
```
Comment (c-bik) [inline src/smpp.erl]: Is also not required!

File: src/smpp.erl (hunk 6 above)
Comment (kirankurup) [reply src/smpp.erl]: This is done to address the issue of having spaces in the input HexStream and not having it, to a common place..
e.g: 00 00 00 04
OR
00000004

Comment (c-bik) [reply src/smpp.erl]: Possible combinations:
```erlang
decode("00 00 00 04")
decode("00000004")
decode(<<"00 00 00 04">>)
decode(<<"00000004">>)
```

The following code should take care of all the four possibilities above.
```erlang
decode(HexList) when is_list(HexList) ->
    decode(list_to_binary(HexList));
decode(HexBin) when is_binary(HexBin) ->
    Hex = re:replace(HexBin, " ", "", [global, {return, binary}]),
...
```

This is addressing the optional space removal in a common place, isn't it?

File: src/smpp.erl (hunk 5 above)
Comment (kirankurup) [reply src/smpp.erl]: String:right is marked Obsolete

File: src/smpp.erl (hunk 7 above)
Comment (c-bik) [reply src/smpp.erl]: For example!
```erlang
[binary_to_integer(B, 16) || <<B:2/binary>> <= <<"000A0F">>].
[0,10,15]
```

**Pull request merged**